File: internal/repository/vat_return_repo.go
package repository

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"time"

	"github.com/zajca/zfaktury/internal/domain"
)

// VATReturnRepository handles persistence of VATReturn entities.
type VATReturnRepository struct {
	db *sql.DB
}

// NewVATReturnRepository creates a new VATReturnRepository.
func NewVATReturnRepository(db *sql.DB) *VATReturnRepository {
	return &VATReturnRepository{db: db}
}

// vatReturnColumns is the list of columns to select for a VATReturn row.
const vatReturnColumns = `id, year, month, quarter, filing_type,
	output_vat_base_21, output_vat_amount_21, output_vat_base_12, output_vat_amount_12, output_vat_base_0,
	reverse_charge_base_21, reverse_charge_amount_21, reverse_charge_base_12, reverse_charge_amount_12,
	input_vat_base_21, input_vat_amount_21, input_vat_base_12, input_vat_amount_12,
	total_output_vat, total_input_vat, net_vat,
	xml_data, status, filed_at, created_at, updated_at`

// scanVATReturn scans a VATReturn from a row.
func scanVATReturn(s scanner) (*domain.VATReturn, error) {
	vr := &domain.VATReturn{}
	var filedAtStr sql.NullString
	var createdAtStr, updatedAtStr string
	var xmlData []byte

	err := s.Scan(
		&vr.ID, &vr.Period.Year, &vr.Period.Month, &vr.Period.Quarter, &vr.FilingType,
		&vr.OutputVATBase21, &vr.OutputVATAmount21, &vr.OutputVATBase12, &vr.OutputVATAmount12, &vr.OutputVATBase0,
		&vr.ReverseChargeBase21, &vr.ReverseChargeAmount21, &vr.ReverseChargeBase12, &vr.ReverseChargeAmount12,
		&vr.InputVATBase21, &vr.InputVATAmount21, &vr.InputVATBase12, &vr.InputVATAmount12,
		&vr.TotalOutputVAT, &vr.TotalInputVAT, &vr.NetVAT,
		&xmlData, &vr.Status, &filedAtStr,
		&createdAtStr, &updatedAtStr,
	)
	if err != nil {
		return nil, err
	}

	vr.XMLData = xmlData

	vr.CreatedAt, err = parseDate(time.RFC3339, createdAtStr)
	if err != nil {
		return nil, fmt.Errorf("scanning vat_return created_at: %w", err)
	}
	vr.UpdatedAt, err = parseDate(time.RFC3339, updatedAtStr)
	if err != nil {
		return nil, fmt.Errorf("scanning vat_return updated_at: %w", err)
	}
	vr.FiledAt, err = parseDatePtr(time.RFC3339, filedAtStr)
	if err != nil {
		return nil, fmt.Errorf("scanning vat_return filed_at: %w", err)
	}

	return vr, nil
}

// Create inserts a new VAT return into the database.
func (r *VATReturnRepository) Create(ctx context.Context, vr *domain.VATReturn) error {
	now := time.Now()
	vr.CreatedAt = now
	vr.UpdatedAt = now
	if vr.Status == "" {
		vr.Status = domain.FilingStatusDraft
	}

	result, err := r.db.ExecContext(ctx, `
		INSERT INTO vat_returns (
			year, month, quarter, filing_type,
			output_vat_base_21, output_vat_amount_21, output_vat_base_12, output_vat_amount_12, output_vat_base_0,
			reverse_charge_base_21, reverse_charge_amount_21, reverse_charge_base_12, reverse_charge_amount_12,
			input_vat_base_21, input_vat_amount_21, input_vat_base_12, input_vat_amount_12,
			total_output_vat, total_input_vat, net_vat,
			xml_data, status, filed_at, created_at, updated_at
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		vr.Period.Year, vr.Period.Month, vr.Period.Quarter, vr.FilingType,
		vr.OutputVATBase21, vr.OutputVATAmount21, vr.OutputVATBase12, vr.OutputVATAmount12, vr.OutputVATBase0,
		vr.ReverseChargeBase21, vr.ReverseChargeAmount21, vr.ReverseChargeBase12, vr.ReverseChargeAmount12,
		vr.InputVATBase21, vr.InputVATAmount21, vr.InputVATBase12, vr.InputVATAmount12,
		vr.TotalOutputVAT, vr.TotalInputVAT, vr.NetVAT,
		nil, vr.Status, nil,
		vr.CreatedAt.Format(time.RFC3339), vr.UpdatedAt.Format(time.RFC3339),
	)
	if err != nil {
		return fmt.Errorf("inserting vat_return: %w", err)
	}

	id, err := result.LastInsertId()
	if err != nil {
		return fmt.Errorf("getting last insert id for vat_return: %w", err)
	}
	vr.ID = id
	return nil
}

// Update modifies an existing VAT return.
func (r *VATReturnRepository) Update(ctx context.Context, vr *domain.VATReturn) error {
	vr.UpdatedAt = time.Now()

	var filedAt any
	if vr.FiledAt != nil {
		filedAt = vr.FiledAt.Format(time.RFC3339)
	}

	_, err := r.db.ExecContext(ctx, `
		UPDATE vat_returns SET
			year = ?, month = ?, quarter = ?, filing_type = ?,
			output_vat_base_21 = ?, output_vat_amount_21 = ?,
			output_vat_base_12 = ?, output_vat_amount_12 = ?,
			output_vat_base_0 = ?,
			reverse_charge_base_21 = ?, reverse_charge_amount_21 = ?,
			reverse_charge_base_12 = ?, reverse_charge_amount_12 = ?,
			input_vat_base_21 = ?, input_vat_amount_21 = ?,
			input_vat_base_12 = ?, input_vat_amount_12 = ?,
			total_output_vat = ?, total_input_vat = ?, net_vat = ?,
			xml_data = ?, status = ?, filed_at = ?, updated_at = ?
		WHERE id = ?`,
		vr.Period.Year, vr.Period.Month, vr.Period.Quarter, vr.FilingType,
		vr.OutputVATBase21, vr.OutputVATAmount21,
		vr.OutputVATBase12, vr.OutputVATAmount12,
		vr.OutputVATBase0,
		vr.ReverseChargeBase21, vr.ReverseChargeAmount21,
		vr.ReverseChargeBase12, vr.ReverseChargeAmount12,
		vr.InputVATBase21, vr.InputVATAmount21,
		vr.InputVATBase12, vr.InputVATAmount12,
		vr.TotalOutputVAT, vr.TotalInputVAT, vr.NetVAT,
		vr.XMLData, vr.Status, filedAt,
		vr.UpdatedAt.Format(time.RFC3339), vr.ID,
	)
	if err != nil {
		return fmt.Errorf("updating vat_return %d: %w", vr.ID, err)
	}
	return nil
}

// Delete removes a VAT return by ID.
func (r *VATReturnRepository) Delete(ctx context.Context, id int64) error {
	result, err := r.db.ExecContext(ctx, `DELETE FROM vat_returns WHERE id = ?`, id)
	if err != nil {
		return fmt.Errorf("deleting vat_return %d: %w", id, err)
	}

	rows, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("checking rows affected for vat_return %d delete: %w", id, err)
	}
	if rows == 0 {
		return fmt.Errorf("vat_return %d: %w", id, domain.ErrNotFound)
	}
	return nil
}

// GetByID retrieves a VAT return by its ID.
func (r *VATReturnRepository) GetByID(ctx context.Context, id int64) (*domain.VATReturn, error) {
	row := r.db.QueryRowContext(ctx, `SELECT `+vatReturnColumns+` FROM vat_returns WHERE id = ?`, id)
	vr, err := scanVATReturn(row)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, fmt.Errorf("vat_return %d: %w", id, domain.ErrNotFound)
		}
		return nil, fmt.Errorf("querying vat_return %d: %w", id, err)
	}
	return vr, nil
}

// List retrieves all VAT returns for a given year.
func (r *VATReturnRepository) List(ctx context.Context, year int) ([]domain.VATReturn, error) {
	rows, err := r.db.QueryContext(ctx,
		`SELECT `+vatReturnColumns+` FROM vat_returns WHERE year = ? ORDER BY month ASC, quarter ASC`,
		year,
	)
	if err != nil {
		return nil, fmt.Errorf("listing vat_returns for year %d: %w", year, err)
	}
	defer rows.Close()

	var result []domain.VATReturn
	for rows.Next() {
		vr, err := scanVATReturn(rows)
		if err != nil {
			return nil, fmt.Errorf("scanning vat_return row: %w", err)
		}
		result = append(result, *vr)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating vat_return rows: %w", err)
	}
	return result, nil
}

// GetByPeriod retrieves a VAT return for a specific period.
func (r *VATReturnRepository) GetByPeriod(ctx context.Context, year, month, quarter int, filingType string) (*domain.VATReturn, error) {
	row := r.db.QueryRowContext(ctx,
		`SELECT `+vatReturnColumns+` FROM vat_returns WHERE year = ? AND month = ? AND quarter = ? AND filing_type = ?`,
		year, month, quarter, filingType,
	)
	vr, err := scanVATReturn(row)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, fmt.Errorf("vat_return for period %d/%d/Q%d (%s): %w", year, month, quarter, filingType, domain.ErrNotFound)
		}
		return nil, fmt.Errorf("querying vat_return by period: %w", err)
	}
	return vr, nil
}

// LinkInvoices associates invoices with a VAT return via the junction table.
func (r *VATReturnRepository) LinkInvoices(ctx context.Context, vatReturnID int64, invoiceIDs []int64) error {
	tx, err := r.db.BeginTx(ctx, nil)
	if err != nil {
		return fmt.Errorf("beginning transaction for linking invoices: %w", err)
	}
	defer tx.Rollback()

	_, err = tx.ExecContext(ctx, `DELETE FROM vat_return_invoices WHERE vat_return_id = ?`, vatReturnID)
	if err != nil {
		return fmt.Errorf("clearing existing invoice links for vat_return %d: %w", vatReturnID, err)
	}

	for _, invID := range invoiceIDs {
		_, err = tx.ExecContext(ctx,
			`INSERT INTO vat_return_invoices (vat_return_id, invoice_id) VALUES (?, ?)`,
			vatReturnID, invID,
		)
		if err != nil {
			return fmt.Errorf("linking invoice %d to vat_return %d: %w", invID, vatReturnID, err)
		}
	}

	if err := tx.Commit(); err != nil {
		return fmt.Errorf("committing invoice links for vat_return %d: %w", vatReturnID, err)
	}
	return nil
}

// LinkExpenses associates expenses with a VAT return via the junction table.
func (r *VATReturnRepository) LinkExpenses(ctx context.Context, vatReturnID int64, expenseIDs []int64) error {
	tx, err := r.db.BeginTx(ctx, nil)
	if err != nil {
		return fmt.Errorf("beginning transaction for linking expenses: %w", err)
	}
	defer tx.Rollback()

	_, err = tx.ExecContext(ctx, `DELETE FROM vat_return_expenses WHERE vat_return_id = ?`, vatReturnID)
	if err != nil {
		return fmt.Errorf("clearing existing expense links for vat_return %d: %w", vatReturnID, err)
	}

	for _, expID := range expenseIDs {
		_, err = tx.ExecContext(ctx,
			`INSERT INTO vat_return_expenses (vat_return_id, expense_id) VALUES (?, ?)`,
			vatReturnID, expID,
		)
		if err != nil {
			return fmt.Errorf("linking expense %d to vat_return %d: %w", expID, vatReturnID, err)
		}
	}

	if err := tx.Commit(); err != nil {
		return fmt.Errorf("committing expense links for vat_return %d: %w", vatReturnID, err)
	}
	return nil
}

// GetLinkedInvoiceIDs returns invoice IDs linked to a VAT return.
func (r *VATReturnRepository) GetLinkedInvoiceIDs(ctx context.Context, vatReturnID int64) ([]int64, error) {
	rows, err := r.db.QueryContext(ctx,
		`SELECT invoice_id FROM vat_return_invoices WHERE vat_return_id = ? ORDER BY invoice_id`,
		vatReturnID,
	)
	if err != nil {
		return nil, fmt.Errorf("querying linked invoices for vat_return %d: %w", vatReturnID, err)
	}
	defer rows.Close()

	var ids []int64
	for rows.Next() {
		var id int64
		if err := rows.Scan(&id); err != nil {
			return nil, fmt.Errorf("scanning linked invoice id: %w", err)
		}
		ids = append(ids, id)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating linked invoice ids: %w", err)
	}
	return ids, nil
}

// GetLinkedExpenseIDs returns expense IDs linked to a VAT return.
func (r *VATReturnRepository) GetLinkedExpenseIDs(ctx context.Context, vatReturnID int64) ([]int64, error) {
	rows, err := r.db.QueryContext(ctx,
		`SELECT expense_id FROM vat_return_expenses WHERE vat_return_id = ? ORDER BY expense_id`,
		vatReturnID,
	)
	if err != nil {
		return nil, fmt.Errorf("querying linked expenses for vat_return %d: %w", vatReturnID, err)
	}
	defer rows.Close()

	var ids []int64
	for rows.Next() {
		var id int64
		if err := rows.Scan(&id); err != nil {
			return nil, fmt.Errorf("scanning linked expense id: %w", err)
		}
		ids = append(ids, id)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating linked expense ids: %w", err)
	}
	return ids, nil
}

File: rust/crates/zfaktury-db/src/repos/vat_return_repo.rs
use crate::helpers::*;
use chrono::Local;
use rusqlite::{Connection, Row, params};
use std::sync::Mutex;
use zfaktury_core::repository::traits::VATReturnRepo;
use zfaktury_domain::*;

pub struct SqliteVATReturnRepo {
    conn: Mutex<Connection>,
}
impl SqliteVATReturnRepo {
    pub fn new(conn: Connection) -> Self {
        Self {
            conn: Mutex::new(conn),
        }
    }
}

fn parse_ft(s: &str) -> FilingType {
    match s {
        "corrective" => FilingType::Corrective,
        "supplementary" => FilingType::Supplementary,
        _ => FilingType::Regular,
    }
}
fn parse_fs(s: &str) -> FilingStatus {
    match s {
        "ready" => FilingStatus::Ready,
        "filed" => FilingStatus::Filed,
        _ => FilingStatus::Draft,
    }
}

fn scan(row: &Row<'_>) -> rusqlite::Result<VATReturn> {
    let ft: String = row.get("filing_type")?;
    let st: String = row.get("status")?;
    let fa: Option<String> = row.get("filed_at")?;
    let c: String = row.get("created_at")?;
    let u: String = row.get("updated_at")?;
    Ok(VATReturn {
        id: row.get("id")?,
        period: TaxPeriod {
            year: row.get("year")?,
            month: row.get("month")?,
            quarter: row.get("quarter")?,
        },
        filing_type: parse_ft(&ft),
        output_vat_base_21: Amount::from_halere(row.get::<_, i64>("output_vat_base_21")?),
        output_vat_amount_21: Amount::from_halere(row.get::<_, i64>("output_vat_amount_21")?),
        output_vat_base_12: Amount::from_halere(row.get::<_, i64>("output_vat_base_12")?),
        output_vat_amount_12: Amount::from_halere(row.get::<_, i64>("output_vat_amount_12")?),
        output_vat_base_0: Amount::from_halere(row.get::<_, i64>("output_vat_base_0")?),
        reverse_charge_base_21: Amount::from_halere(row.get::<_, i64>("reverse_charge_base_21")?),
        reverse_charge_amount_21: Amount::from_halere(
            row.get::<_, i64>("reverse_charge_amount_21")?,
        ),
        reverse_charge_base_12: Amount::from_halere(row.get::<_, i64>("reverse_charge_base_12")?),
        reverse_charge_amount_12: Amount::from_halere(
            row.get::<_, i64>("reverse_charge_amount_12")?,
        ),
        input_vat_base_21: Amount::from_halere(row.get::<_, i64>("input_vat_base_21")?),
        input_vat_amount_21: Amount::from_halere(row.get::<_, i64>("input_vat_amount_21")?),
        input_vat_base_12: Amount::from_halere(row.get::<_, i64>("input_vat_base_12")?),
        input_vat_amount_12: Amount::from_halere(row.get::<_, i64>("input_vat_amount_12")?),
        total_output_vat: Amount::from_halere(row.get::<_, i64>("total_output_vat")?),
        total_input_vat: Amount::from_halere(row.get::<_, i64>("total_input_vat")?),
        net_vat: Amount::from_halere(row.get::<_, i64>("net_vat")?),
        xml_data: row
            .get::<_, Option<Vec<u8>>>("xml_data")?
            .unwrap_or_default(),
        status: parse_fs(&st),
        filed_at: parse_datetime_optional(fa.as_deref()).unwrap_or(None),
        created_at: parse_datetime(&c).unwrap_or_default(),
        updated_at: parse_datetime(&u).unwrap_or_default(),
    })
}

impl VATReturnRepo for SqliteVATReturnRepo {
    fn create(&self, vr: &mut VATReturn) -> Result<(), DomainError> {
        let conn = self.conn.lock().unwrap();
        let now = format_datetime(&Local::now().naive_local());
        conn.execute("INSERT INTO vat_returns (year,month,quarter,filing_type,output_vat_base_21,output_vat_amount_21,output_vat_base_12,output_vat_amount_12,output_vat_base_0,reverse_charge_base_21,reverse_charge_amount_21,reverse_charge_base_12,reverse_charge_amount_12,input_vat_base_21,input_vat_amount_21,input_vat_base_12,input_vat_amount_12,total_output_vat,total_input_vat,net_vat,xml_data,status,filed_at,created_at,updated_at) VALUES (?1,?2,?3,?4,?5,?6,?7,?8,?9,?10,?11,?12,?13,?14,?15,?16,?17,?18,?19,?20,?21,?22,?23,?24,?24)",
            params![vr.period.year,vr.period.month,vr.period.quarter,vr.filing_type.to_string(),vr.output_vat_base_21.halere(),vr.output_vat_amount_21.halere(),vr.output_vat_base_12.halere(),vr.output_vat_amount_12.halere(),vr.output_vat_base_0.halere(),vr.reverse_charge_base_21.halere(),vr.reverse_charge_amount_21.halere(),vr.reverse_charge_base_12.halere(),vr.reverse_charge_amount_12.halere(),vr.input_vat_base_21.halere(),vr.input_vat_amount_21.halere(),vr.input_vat_base_12.halere(),vr.input_vat_amount_12.halere(),vr.total_output_vat.halere(),vr.total_input_vat.halere(),vr.net_vat.halere(),&vr.xml_data as &[u8],vr.status.to_string(),format_datetime_opt(&vr.filed_at),now])
            .map_err(|e|{log::error!("insert vat return: {e}");DomainError::InvalidInput})?;
        vr.id = conn.last_insert_rowid();
        Ok(())
    }
    fn update(&self, vr: &mut VATReturn) -> Result<(), DomainError> {
        let conn = self.conn.lock().unwrap();
        let now = format_datetime(&Local::now().naive_local());
        conn.execute("UPDATE vat_returns SET output_vat_base_21=?1,output_vat_amount_21=?2,output_vat_base_12=?3,output_vat_amount_12=?4,output_vat_base_0=?5,reverse_charge_base_21=?6,reverse_charge_amount_21=?7,reverse_charge_base_12=?8,reverse_charge_amount_12=?9,input_vat_base_21=?10,input_vat_amount_21=?11,input_vat_base_12=?12,input_vat_amount_12=?13,total_output_vat=?14,total_input_vat=?15,net_vat=?16,xml_data=?17,status=?18,filed_at=?19,updated_at=?20 WHERE id=?21",
            params![vr.output_vat_base_21.halere(),vr.output_vat_amount_21.halere(),vr.output_vat_base_12.halere(),vr.output_vat_amount_12.halere(),vr.output_vat_base_0.halere(),vr.reverse_charge_base_21.halere(),vr.reverse_charge_amount_21.halere(),vr.reverse_charge_base_12.halere(),vr.reverse_charge_amount_12.halere(),vr.input_vat_base_21.halere(),vr.input_vat_amount_21.halere(),vr.input_vat_base_12.halere(),vr.input_vat_amount_12.halere(),vr.total_output_vat.halere(),vr.total_input_vat.halere(),vr.net_vat.halere(),&vr.xml_data as &[u8],vr.status.to_string(),format_datetime_opt(&vr.filed_at),now,vr.id])
            .map_err(|e|{log::error!("update vat return: {e}");DomainError::InvalidInput})?;
        Ok(())
    }
    fn delete(&self, id: i64) -> Result<(), DomainError> {
        let conn = self.conn.lock().unwrap();
        let r = conn
            .execute("DELETE FROM vat_returns WHERE id=?1", params![id])
            .map_err(|e| {
                log::error!("del: {e}");
                DomainError::InvalidInput
            })?;
        if r == 0 {
            return Err(DomainError::NotFound);
        }
        Ok(())
    }
    fn get_by_id(&self, id: i64) -> Result<VATReturn, DomainError> {
        let conn = self.conn.lock().unwrap();
        conn.query_row("SELECT * FROM vat_returns WHERE id=?1", params![id], scan)
            .map_err(|e| match e {
                rusqlite::Error::QueryReturnedNoRows => DomainError::NotFound,
                _ => {
                    log::error!("q: {e}");
                    DomainError::InvalidInput
                }
            })
    }
    fn list(&self, year: i32) -> Result<Vec<VATReturn>, DomainError> {
        let conn = self.conn.lock().unwrap();
        let mut s = conn
            .prepare("SELECT * FROM vat_returns WHERE year=?1 ORDER BY month,quarter")
            .map_err(|e| {
                log::error!("prep: {e}");
                DomainError::InvalidInput
            })?;
        s.query_map(params![year], scan)
            .map_err(|e| {
                log::error!("list: {e}");
                DomainError::InvalidInput
            })?
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| {
                log::error!("scan: {e}");
                DomainError::InvalidInput
            })
    }
    fn get_by_period(
        &self,
        year: i32,
        month: i32,
        quarter: i32,
        filing_type: &str,
    ) -> Result<VATReturn, DomainError> {
        let conn = self.conn.lock().unwrap();
        conn.query_row("SELECT * FROM vat_returns WHERE year=?1 AND month=?2 AND quarter=?3 AND filing_type=?4",params![year,month,quarter,filing_type],scan)
            .map_err(|e| match e { rusqlite::Error::QueryReturnedNoRows=>DomainError::NotFound, _=>{log::error!("q: {e}");DomainError::InvalidInput}})
    }
    fn link_invoices(&self, vat_return_id: i64, invoice_ids: &[i64]) -> Result<(), DomainError> {
        let conn = self.conn.lock().unwrap();
        for id in invoice_ids {
            conn.execute("INSERT OR IGNORE INTO vat_return_invoices (vat_return_id,invoice_id) VALUES (?1,?2)",params![vat_return_id,id]).map_err(|e|{log::error!("link inv: {e}");DomainError::InvalidInput})?;
        }
        Ok(())
    }
    fn link_expenses(&self, vat_return_id: i64, expense_ids: &[i64]) -> Result<(), DomainError> {
        let conn = self.conn.lock().unwrap();
        for id in expense_ids {
            conn.execute("INSERT OR IGNORE INTO vat_return_expenses (vat_return_id,expense_id) VALUES (?1,?2)",params![vat_return_id,id]).map_err(|e|{log::error!("link exp: {e}");DomainError::InvalidInput})?;
        }
        Ok(())
    }
    fn get_linked_invoice_ids(&self, vat_return_id: i64) -> Result<Vec<i64>, DomainError> {
        let conn = self.conn.lock().unwrap();
        let mut s = conn
            .prepare("SELECT invoice_id FROM vat_return_invoices WHERE vat_return_id=?1")
            .map_err(|e| {
                log::error!("prep: {e}");
                DomainError::InvalidInput
            })?;
        s.query_map(params![vat_return_id], |r| r.get(0))
            .map_err(|e| {
                log::error!("list: {e}");
                DomainError::InvalidInput
            })?
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| {
                log::error!("scan: {e}");
                DomainError::InvalidInput
            })
    }
    fn get_linked_expense_ids(&self, vat_return_id: i64) -> Result<Vec<i64>, DomainError> {
        let conn = self.conn.lock().unwrap();
        let mut s = conn
            .prepare("SELECT expense_id FROM vat_return_expenses WHERE vat_return_id=?1")
            .map_err(|e| {
                log::error!("prep: {e}");
                DomainError::InvalidInput
            })?;
        s.query_map(params![vat_return_id], |r| r.get(0))
            .map_err(|e| {
                log::error!("list: {e}");
                DomainError::InvalidInput
            })?
            .collect::<Result<Vec<_>, _>>()
            .map_err(|e| {
                log::error!("scan: {e}");
                DomainError::InvalidInput
            })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_db::new_test_db;
    #[test]
    fn test_create_and_get() {
        let conn = new_test_db();
        let repo = SqliteVATReturnRepo::new(conn);
        let mut vr = VATReturn {
            id: 0,
            period: TaxPeriod {
                year: 2025,
                month: 1,
                quarter: 0,
            },
            filing_type: FilingType::Regular,
            output_vat_base_21: Amount::ZERO,
            output_vat_amount_21: Amount::ZERO,
            output_vat_base_12: Amount::ZERO,
            output_vat_amount_12: Amount::ZERO,
            output_vat_base_0: Amount::ZERO,
            reverse_charge_base_21: Amount::ZERO,
            reverse_charge_amount_21: Amount::ZERO,
            reverse_charge_base_12: Amount::ZERO,
            reverse_charge_amount_12: Amount::ZERO,
            input_vat_base_21: Amount::ZERO,
            input_vat_amount_21: Amount::ZERO,
            input_vat_base_12: Amount::ZERO,
            input_vat_amount_12: Amount::ZERO,
            total_output_vat: Amount::ZERO,
            total_input_vat: Amount::ZERO,
            net_vat: Amount::ZERO,
            xml_data: Vec::new(),
            status: FilingStatus::Draft,
            filed_at: None,
            created_at: Default::default(),
            updated_at: Default::default(),
        };
        repo.create(&mut vr).unwrap();
        assert!(vr.id > 0);
        let f = repo.get_by_id(vr.id).unwrap();
        assert_eq!(f.period.year, 2025);
    }
}
